lib/dal/src/action/dependency_graph.rs

```diff
@@ -122,11 +129,19 @@ impl ActionDependencyGraph {
                     // components.
                     component_dependencies.update_edge(source_component_index, component_index, ());
                 }
+                new_components_to_process.insert(inferred_connection.source_component_id);
+            }
+            // only add new components to the queue if they are not already in the queue
+            for component_id in new_components_to_process {
+                if !components_to_process.contains(&component_id) {
+                    components_to_process.push_back(component_id);
+                }
             }
```

Comment:
Rather than checking if it's currently in the queue, it would probably be better to keep track of which ones have already been "Seen" (already visited, and when adding to the queue), and only add if it isn't in the seen set. Only checking if it's currently in the queue could potentially end up in an infinite loop if there's a cycle.